File: framework/backend/python/packages/k1s0-consensus/src/k1s0_consensus/fencing.py
"""Fence token validation for distributed lock safety."""

from __future__ import annotations

import threading


class FencingValidator:
    """Thread-safe monotonic fence token validator.

    A fence token is a monotonically increasing integer issued when a lock
    is acquired. Resources protected by fencing reject operations that
    present a token lower than the highest previously accepted token,
    preventing stale lock holders from corrupting data.

    Example::

        validator = FencingValidator()
        assert validator.validate(1) is True
        assert validator.validate(2) is True
        assert validator.validate(1) is False  # stale token rejected

    This class is thread-safe and can be shared across threads.
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._highest_token: int = -1

    def validate(self, token: int) -> bool:
        """Validate a fence token against the highest seen token.

        The token is accepted only if it is strictly greater than the
        highest previously accepted token. On acceptance, the highest
        token is updated.

        Args:
            token: The fence token to validate.

        Returns:
            True if the token is valid (higher than any previously seen),
            False if stale.
        """
        with self._lock:
            if token > self._highest_token:
                self._highest_token = token
                return True
            return False

    @property
    def highest_token(self) -> int:
        """The highest fence token accepted so far.

        Returns -1 if no token has been validated yet.
        """
        with self._lock:
            return self._highest_token

    def reset(self) -> None:
        """Reset the validator to its initial state.

        This is primarily useful in testing. In production, resetting
        the validator can lead to stale token acceptance.
        """
        with self._lock:
            self._highest_token = -1

File: framework/backend/go/k1s0-consensus/fencing.go
package consensus

import (
	"fmt"
	"sync/atomic"
)

// FencingValidator validates that fence tokens are monotonically increasing.
// It is used to detect stale lock holders that may still attempt to perform
// operations after their lock has expired and been re-acquired by another node.
//
// Usage:
//
//	validator := consensus.NewFencingValidator()
//	if !validator.Validate(guard.FenceToken) {
//	    // Stale operation detected, reject it.
//	}
type FencingValidator struct {
	highestToken atomic.Uint64
}

// NewFencingValidator creates a new FencingValidator.
func NewFencingValidator() *FencingValidator {
	return &FencingValidator{}
}

// Validate checks that the given token is strictly greater than any
// previously validated token. Returns true if the token is valid
// (monotonically increasing), false otherwise.
func (v *FencingValidator) Validate(token uint64) bool {
	for {
		current := v.highestToken.Load()
		if token <= current {
			metricsFenceTokenViolations.Inc()
			return false
		}
		if v.highestToken.CompareAndSwap(current, token) {
			return true
		}
		// CAS failed, another goroutine updated; retry.
	}
}

// ValidateOrError is like Validate but returns an error on failure.
func (v *FencingValidator) ValidateOrError(token uint64) error {
	if !v.Validate(token) {
		return fmt.Errorf("consensus: token %d is not greater than current: %w",
			token, ErrFenceTokenViolation)
	}
	return nil
}

// Current returns the highest validated token seen so far.
func (v *FencingValidator) Current() uint64 {
	return v.highestToken.Load()
}

// Reset resets the validator to its initial state (token 0).
func (v *FencingValidator) Reset() {
	v.highestToken.Store(0)
}
